lib/extensions/explain_validity.py

```diff
@@ -0,0 +1,154 @@
+
+from shapely.validation import explain_validity
+
+import inkex
+
+from ..commands import get_command_description
+from ..i18n import _
+from ..svg import get_correction_transform
+from ..svg.tags import (INKSCAPE_GROUPMODE, INKSCAPE_LABEL,
+                        SODIPODI_INSENSITIVE, SODIPODI_ROLE, SVG_GROUP_TAG,
+                        SVG_PATH_TAG, SVG_TEXT_TAG, SVG_TSPAN_TAG, SVG_USE_TAG,
+                        XLINK_HREF)
+from .base import InkstitchExtension
+from .commands import CommandsExtension
+
+
+class ExplainValidity(InkstitchExtension):
+    def effect(self):
+        if not self.get_elements():
+            return
+
+        valid_shapes = True
+        for element in self.elements:
+            shape = element.shape
+            if not shape.is_valid:
+                valid_shapes = False
+                self.insert_invalid_pointer(shape)
+
+        if valid_shapes is True:
+            inkex.errormsg(_("All selected shapes are valid!"))
+
+    def insert_invalid_pointer(self, shape):
+        layer = self.get_or_create_validity_layer()
+        correction_transform = get_correction_transform(layer, child=True)
+
+        shapely_msg = explain_validity(shape)
+        message, point_x, point_y = re.findall(r".+?(?=\[)|\d+\.\d+", shapely_msg)
+
+        path = inkex.etree.Element(
+            SVG_PATH_TAG,
+            {
+                "id": self.uniqueId("inkstitch__invalid_pointer__"),
+                "d": "m %s,%s 4,20 h -8 l 4,-20" % (point_x, point_y),
+                "style": "fill:#ff0000;stroke:#ffffff;stroke-width:0.2;",
+                INKSCAPE_LABEL: _('Invalid Pointer'),
+                "transform": correction_transform
+             }
+             )
+        layer.insert(0, path)
+
+        text = inkex.etree.Element(
+            SVG_TEXT_TAG,
+            {
+                "x": point_x,
+                "y": str(float(point_y) + 30),
+                "transform": correction_transform,
+                "style": "fill:#ff0000;troke:#ffffff;stroke-width:0.2;font-size:8px;text-align:center;text-anchor:middle"
+            }
+            )
+        layer.append(text)
+
+        tspan = inkex.etree.Element(SVG_TSPAN_TAG)
+        if message == "Self-intersection":
+            tspan.text = _("Self-intersection")
+        elif message == "Hole lies outside shell":
+            tspan.text = _("Gap")
+        else:
+            # different error (message not translatable)
+            tspan.text = message
+        text.append(tspan)
+
+    def get_or_create_validity_layer(self):
+        svg = self.document.getroot()
+        layer = svg.find(".//*[@id='__validity_layer__']")
+
+        if layer is None:
+            layer = inkex.etree.Element(
+                SVG_GROUP_TAG,
+                {
+                    'id': '__validity_layer__',
+                    INKSCAPE_LABEL: _('Explain Validity'),
+                    INKSCAPE_GROUPMODE: 'layer',
+                    SODIPODI_INSENSITIVE: "true"
```

Comment:
I'd love to hear your thinking behind locking the validity layer.  What problems does that solve for the user?